torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+// results into shared memory or local registers before moving to global memory.
+// Analogous to TVM Cache_Read
+TensorView* TensorView::cache_after() {
+  FusionGuard fg(this->fusion());
+
+  // Get all the uses for this Tensorview
+  TORCH_CHECK(
+      !this->fusion()->hasOutput(this),
+      "Error adding cache_after ",
+      this,
+      " we restrict using cache_after on an output.");
+
+  // Create Consumer Domain
+  // Keep Broadcast Axis (Permanent)
+  auto root_domain = getRootDomain();
+  std::vector<IterDomain*> new_root_domain;
+  for (auto root : root_domain) {
+    if (root->isBroadcast()) {
+      new_root_domain.push_back(new IterDomain(
+          root->start(),
+          root->extent(),
+          root->parallel_method(),
+          false,
+          false,
+          true));
+    } else if (!root->isBroadcast() && !root->isReduction()) {
+      new_root_domain.push_back(new IterDomain(
+          root->start(), root->extent(), root->parallel_method()));
+    }
+  }
+
+  // This domain will be the producer, so create the consumer
+  TensorView* consumer = new TensorView(
+      new TensorDomain(new_root_domain), this->getDataType().value());
+
+  // Set domain of producer - No Change
+  TensorView* producer = this;
+
+  // Insert consumer - Cache_After (CA) - after this TV.
+  // Before: This TV -> [Use Op] -> Next TV
+  // After:  This TV -> [Set Op] -> New CA TV -> [Use Op] -> Next TV
+
+  // Expr* consumer_uses =
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
+  for (auto expr : uses_expr) {
+    createExprProducer(expr, this, consumer);
+  }
+
+  // Expr* consumer_origin =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // Before: This TV -> Next TV
+  // After:  This TV -> New TV (After) -> Next TV
+  if (this->hasComputeAt()) {
+    TransformReplay::replayCasP(consumer, producer, -1);
+
+    auto rel_ca_pos = this->getRelativeComputeAtAxis();
+    auto this_ca_pos = this->getThisComputeAtAxis();
+    auto this_ca_view = this->getComputeAtView();
+
+    this->computeAt(consumer, this_ca_pos);
+    consumer->setComputeAt(this_ca_view, rel_ca_pos);
+  } else {
+    // Check users of this TV for computeAt for cache_after on inputs
+    std::unordered_set<Expr*> consumer_uses =
+        this->fusion()->unordered_uses(consumer);
+    for (auto expr : consumer_uses) {
+      auto expr_outputs = expr->outputs();
+      for (Val* v : expr_outputs) {
+        if (v->getValType().value() == ValType::TensorView) {
+          TensorView* output = dynamic_cast<TensorView*>(v);
+          if (output->hasComputeAt()) {
+            TransformReplay::replayPasC(consumer, output, -1);
+            auto output_ca_pos = output->getThisComputeAtAxis();
+            consumer->setComputeAt(output, output_ca_pos);
+          }
+        }
+      }
+    }
+  }
+
+  return consumer;
+}
+
+// Create New Expr given consumer - [output of the expression]
+struct CreateExprConsumer : public OptInDispatch {
```

Comment:
Can you put this inside an anonymous namespace?

Reply:
Done